arrow/arrowhead.pyx

```diff
@@ -0,0 +1,149 @@
+from libc.stdint cimport int64_t
+from libc.string cimport memset, memcpy
+from libc.stdlib cimport free
+
+import numpy as np
+cimport numpy as np
+
+from . cimport mersenne
+from . cimport obsidian
+
+
+np.import_array()  # Initialize numpy's C API so it won't segfault
+
+
+cdef class Arrowhead:
+    """Cython interface to the C-coded Gillespie algorithm "obsidian", leaving
+    obsidian.c free of Python dependencies.
+    """
+
+    cdef:
+        int random_seed
+        obsidian.Info info
+        tuple refs
+
+    # __cinit__() is always called immediately on construction, before CPython
+    # considers calling __init__(), which it can skip. `self` is not yet fully
+    # constructed so *BE CAREFUL* invoking Python operations that might touch it!
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    def __cinit__(self, *args):
+        memset(&self.info, 0, sizeof(obsidian.Info))
+        self.info.random_state = <mersenne.MTState*> PyMem_Malloc(sizeof(mersenne.MTState))
+        if not self.info.random_state:
+            raise MemoryError()
+
+    # This will accept any type of array in C-contiguous layout that support
+    # the memoryview protocol (numpy array, array.array, Cython array, buffer
+    # interface) with the declared element type and number of dimensions.
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    def __init__(
+            self,
+            int random_seed,
+            int64_t[:, ::1] stoichiometry not None,
+            double[::1] rates not None,
+
+            int64_t[::1] reactants_lengths not None,
+            int64_t[::1] reactants_indexes not None,
+            int64_t[::1] reactants not None,
+            int64_t[::1] reactions not None,
+
+            int64_t[::1] dependencies_lengths not None,
+            int64_t[::1] dependencies_indexes not None,
+            int64_t[::1] dependencies not None,
+
+            int64_t[::1] substrates_lengths not None,
+            int64_t[::1] substrates_indexes not None,
+            int64_t[::1] substrates not None,
+            *args):
+        self.random_seed = random_seed
+        mersenne.seed(self.info.random_state, random_seed)
+
+        self.info.reactions_count = stoichiometry.shape[0]
+        self.info.substrates_count = stoichiometry.shape[1]
+        self.info.stoichiometry = &stoichiometry[0, 0]
+        self.info.rates = &rates[0]
+
+        self.info.reactants_lengths = &reactants_lengths[0]
+        self.info.reactants_indexes = &reactants_indexes[0]
+        self.info.reactants = &reactants[0]
+        self.info.reactions = &reactions[0]
+
+        self.info.dependencies_lengths = &dependencies_lengths[0]
+        self.info.dependencies_indexes = &dependencies_indexes[0]
+        self.info.dependencies = &dependencies[0]
+
+        self.info.substrates_lengths = &substrates_lengths[0]
+        self.info.substrates_indexes = &substrates_indexes[0]
+        self.info.substrates = &substrates[0]
```

Comment:
Nice, so this `info` struct saves us copying then? Looks like a good improvement.

Reply:
Yes. And we could go further: In obsidian.c, `evolve()` could use those info fields directly rather than copying them all into local variables, but it wouldn't save much.